Type annotations needed error on definition site

Hello

After upgrading nightly rustc, I'm getting a lot of type inference errors on code that was compiling fine previously. Moreover, these are not at the call site, but at the definition site, so there doesn't seem to be any place to put any type annotations, like this:

```
    --> src/spirit.rs:996:1
     |
996  | / impl<O, C> SpiritBuilder for Builder<O, C>
997  | | where
998  | |     Self::Config: DeserializeOwned + Send + Sync + 'static,
999  | |     Self::Opts: StructOpt + Sync + Send + 'static,
...    |
1108 | |     }
1109 | | }
     | |_^ cannot infer type for type parameter `C`
```

It spills more kinds of errors that make little sense, but I have the impression these will have a common cause:

```
error[E0277]: `C` cannot be shared between threads safely
   --> src/spirit.rs:998:45
    |
998 |     Self::Config: DeserializeOwned + Send + Sync + 'static,
    |                                             ^^^^ `C` cannot be shared between threads safely
```

The problematic code is both released on crates.io (crate spirit, version 0.4.16, but I expect that to be happening through most of the history of the crate), and in master (https://github.com/vorner/spirit/commit/1495c1cdef17663e8f847f037abeceb37f310fad).

I think this is a regression, as it was compiling previously.

It compiles fine with older nightly (and stable is fine way back, this particular code is looking mostly the same for some 2 years or more):

```
rustc 1.53.0-nightly (07e0e2ec2 2021-03-24)
```

Breaks with

```
rustc 1.53.0-nightly (6df26f897 2021-04-20)
binary: rustc
commit-hash: 6df26f897cffb2d86880544bb451c6b5f8509b2d
commit-date: 2021-04-20
host: x86_64-unknown-linux-gnu
release: 1.53.0-nightly
LLVM version: 12.0.0
```

(Unfortunately, I haven't been keeping fully up to date recently, so I don't have a smaller range).

--- a/compiler/rustc_trait_selection/src/traits/select/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/select/mod.rs
@@ -1361,7 +1361,17 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             ) => false,
 
             (ParamCandidate(other), ParamCandidate(victim)) => {
-                if other.value == victim.value && victim.constness == Constness::NotConst {
+                let value_same_except_bound_vars = other.value.skip_binder()
+                    == victim.value.skip_binder()
+                    && !other.value.skip_binder().has_escaping_bound_vars();
+                if value_same_except_bound_vars {
+                    // See issue #84398. In short, we can generate multiple ParamCandidates which are
+                    // the same except for unused bound vars. Just pick the current one (the should
+                    // both evaluate to the same answer). This is probably best characterized as a
+                    // "hack", since we might prefer to just do our best to *not* create essentially
+                    // duplicate candidates in the first place.
+                    true
+                } else if other.value == victim.value && victim.constness == Constness::NotConst {
                     // Drop otherwise equivalent non-const candidates in favor of const candidates.
                     true
                 } else {

--- a/compiler/rustc_trait_selection/src/traits/select/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/select/mod.rs
@@ -1366,11 +1366,11 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                     && !other.value.skip_binder().has_escaping_bound_vars();
                 if value_same_except_bound_vars {
                     // See issue #84398. In short, we can generate multiple ParamCandidates which are
-                    // the same except for unused bound vars. Just pick the current one (the should
-                    // both evaluate to the same answer). This is probably best characterized as a
-                    // "hack", since we might prefer to just do our best to *not* create essentially
-                    // duplicate candidates in the first place.
-                    true
+                    // the same except for unused bound vars. Just pick the one with the fewest bound vars
+                    // or the current one if tied (they should both evaluate to the same answer). This is
+                    // probably best characterized as a "hack", since we might prefer to just do our
+                    // best to *not* create essentially duplicate candidates in the first place.
+                    other.value.bound_vars().len() <= victim.value.bound_vars().len()
                 } else if other.value == victim.value && victim.constness == Constness::NotConst {
                     // Drop otherwise equivalent non-const candidates in favor of const candidates.
                     true

--- a/src/test/ui/lifetimes/issue-84398.rs
+++ b/src/test/ui/lifetimes/issue-84398.rs
@@ -1,0 +1,20 @@
+// check-pass
+
+pub trait Deserialize<'de>: Sized {}
+pub trait DeserializeOwned: for<'de> Deserialize<'de> {}
+
+pub trait Extensible {
+    type Config;
+}
+
+// The `C` here generates a `C: Sized` candidate
+pub trait Installer<C> {
+    fn init<B: Extensible<Config = C>>(&mut self) -> ()
+    where
+        // This clause generates a `for<'de> C: Sized` candidate
+        B::Config: DeserializeOwned,
+    {
+    }
+}
+
+fn main() {}

